src/webgpu/api/validation/createBindGroupLayout.spec.ts

```diff
@@ -402,6 +402,118 @@ g.test('max_resources_per_stage,in_pipeline_layout')
     }, newBindingCountsTowardSamePerStageLimit);
   });
 
+const kMaxResourcesPerPipelineCases = kUnitCaseParamsBuilder
+  .combine('maxedType', kBufferBindingTypes)
+  .beginSubcases()
+  .combine('maxedVisibility', kShaderStages)
+  .filter(p => (bufferBindingTypeInfo({ type: p.maxedType }).validStages & p.maxedVisibility) !== 0)
+  .combine('extraType', kBufferBindingTypes)
+  .combine('extraVisibility', kShaderStages)
+  .filter(
+    p => (bufferBindingTypeInfo({ type: p.extraType }).validStages & p.extraVisibility) !== 0
+  );
+
+// One bind group layout can have a maximum number of each type of dynamic buffer per pipeline. Test
+// that the max works, then add one more buffer of the same-or-different type and same-or-different
+// visibility.
+g.test('max_dynamic_buffers_per_pipeline,in_bind_group_layout')
+  .desc(
+    `
+    Test that the maximum number of bindings of a given dynamic buffer type per-pipeline cannot be
+    exceeded in a single bind group layout.
+    - Test each binding dynamic buffer type.
+    - Test that creation of a bind group layout using the maximum number of bindings works.
+    - Test that creation of a bind group layout using the maximum number of bindings + 1 fails.`
+  )
+  .params(kMaxResourcesPerPipelineCases)
+  .fn(t => {
+    const { maxedType, extraType, maxedVisibility, extraVisibility } = t.params;
+    const maxedEntry: BGLEntry = { buffer: { type: maxedType, hasDynamicOffset: true } };
+    const perPipelineLimit = bindingTypeInfo(maxedEntry).perPipelineLimitClass;
+
+    const maxResourceBindings: GPUBindGroupLayoutEntry[] = [];
+    for (let i = 0; i < perPipelineLimit.maxDynamic; i++) {
+      maxResourceBindings.push({
+        binding: i,
+        visibility: maxedVisibility,
+        ...maxedEntry,
+      });
+    }
+
+    const goodDescriptor = { entries: maxResourceBindings };
+
+    // Control
+    t.device.createBindGroupLayout(goodDescriptor);
+
+    // Add an entry that may count towards the same limit. If it does, it should produce a
+    // validation error.
+    const extraEntry: BGLEntry = { buffer: { type: extraType, hasDynamicOffset: true } };
+    const newDescriptor = clone(goodDescriptor);
+    newDescriptor.entries.push({
+      binding: perPipelineLimit.maxDynamic,
+      visibility: extraVisibility,
+      ...extraEntry,
+    });
+
+    const newBindingCountsTowardSamePerPipelineLimit =
+      perPipelineLimit.class === bindingTypeInfo(extraEntry).perPipelineLimitClass.class;
+    t.expectValidationError(() => {
+      t.device.createBindGroupLayout(newDescriptor);
+    }, newBindingCountsTowardSamePerPipelineLimit);
+  });
+
+// One pipeline layout can have a maximum number of each type of dynamic buffer per pipeline. Test
+// that the max works, then add one more bind group layout with a buffer of the same-or-different
+// type and same-or-different visibility.
+g.test('max_dynamic_buffers_per_pipeline,in_pipeline_layout')
```

Comment:
This file is createBindGroupLayout so it's odd to have pipeline layout tests here (I know there are already some). I think we should move all of the limit tests that apply to both createBindGroupLayout and createPipelineLayout to a new file? (and leave behind notes in both createBindGroupLayout and createPipelineLayout's test files)
Can be a followup PR.